src/shell.zig

```diff
@@ -392,152 +392,160 @@ pub fn copy_path(
-///
-/// One intended use-case is sanity-checking that an executable is present, by running
-/// `my-tool --version`.
-pub fn exec_status_ok(shell: *Shell, comptime cmd: []const u8, cmd_args: anytype) !bool {
+/// Runs the given command and returns its output.
+/// If the output is a single line, the final newline is stripped.
+pub fn exec_stdout(shell: *Shell, comptime cmd: []const u8, cmd_args: anytype) ![]const u8 {
     var argv = try Argv.expand(shell.gpa, cmd, cmd_args);
     defer argv.deinit();
 
-    var child = try shell.create_process(argv.slice());
-    const term = try child.spawnAndWait();
-    return switch (term) {
-        .Exited => |code| code == 0,
-        else => false,
-    };
-}
-
-/// Run the command and return its stdout.
-///
-/// Returns an error if the command exists with a non-zero status or a non-empty stderr.
-///
-/// Trims the trailing newline, if any.
-pub fn exec_stdout(shell: *Shell, comptime cmd: []const u8, cmd_args: anytype) ![]const u8 {
-    return shell.exec_stdout_options(.{}, cmd, cmd_args);
+    var captured_stdout: []const u8 = &.{};
+    try exec_inner(shell, argv.slice(), .{
+        .capture_stdout = &captured_stdout,
+    });
+    return captured_stdout;
 }
 
 pub fn exec_stdout_options(
     shell: *Shell,
     options: struct {
         stdin_slice: ?[]const u8 = null,
-        output_bytes_max: usize = 1024 * 1024 * 32,
     },
     comptime cmd: []const u8,
     cmd_args: anytype,
 ) ![]const u8 {
     var argv = try Argv.expand(shell.gpa, cmd, cmd_args);
     defer argv.deinit();
 
-    var child = try shell.create_process(argv.slice());
-    child.stdin_behavior = if (options.stdin_slice == null) .Ignore else .Pipe;
-    child.stdout_behavior = .Pipe;
-    child.stderr_behavior = .Pipe;
+    var captured_stdout: []const u8 = &.{};
+    try exec_inner(shell, argv.slice(), .{
+        .stdin_slice = options.stdin_slice,
+        .capture_stdout = &captured_stdout,
+    });
+    return captured_stdout;
+}
 
-    var stdout = std.ArrayList(u8).init(shell.gpa);
-    var stderr = std.ArrayList(u8).init(shell.gpa);
-    defer {
-        stdout.deinit();
-        stderr.deinit();
+/// Runs the zig compiler.
+pub fn exec_zig(shell: *Shell, comptime cmd: []const u8, cmd_args: anytype) !void {
+    var argv = Argv.init(shell.gpa);
+    defer argv.deinit();
+
+    try argv.append_new_arg("{s}", .{shell.zig_exe.?});
+    try expand_argv(&argv, cmd, cmd_args);
+
+    return shell.exec_inner(argv.slice(), .{});
+}
+
+fn exec_inner(
+    shell: *Shell,
+    argv: []const []const u8,
+    options: struct {
+        stdin_slice: ?[]const u8 = null,
+        capture_stdout: ?*[]const u8 = null, // Optional out parameter.
+        expect_zero_status: bool = true,
+        output_limit_bytes: usize = 128 * 1024 * 1024,
+        timeout_ns: u64 = 10 * std.time.ns_per_min,
+    },
+) !void {
+    const argv_formatted = try std.mem.join(shell.gpa, " ", argv);
```

Comment:
One can write a lot more code to format comand line in a nice way, but I think this will be sufficiently unambigious for our use-case!